[WIP] Fixes #3779: Allow console access through the API

ekohl opened the pull request:
For now this is a proposal which works for VNC, but isn't tested for spice.

ekohl committed:
Fixes #3779: Allow console access through the API

domcleal commented on app/controllers/api/v2/hosts_controller.rb:
```diff
@@ -133,6 +133,21 @@ def resource_scope
+      def console
+        return unless @host.compute_resource
+        @console = @host.compute_resource.console @host.uuid
+        @result = {
```
It'd be better to use a RABL file here to describe the different console types (spice, vnc, log):

```
def console
  return unless @host.compute_resource
  @console = @host.compute_resource.console @host.uuid
  @console[:name] ||= URI.parse(Setting[:foreman_url]).host
  @console = OpenStruct.new @console
end
```

(no `render` needed)

Then in app/views/api/v2/hosts/console.json.rabl reference `object @console`.  I'd then do something like app/views/api/v2/compute_resources/main.json.rabl (partials) to call one of `{spice,vnc,log}.json.rabl` using the `type` attribute as they're going to use different attributes ("log" especially as we actually have the console contents in `@console[:log]`).

As an example of VNC, you can do something like:

```
attributes :proxy_port => :port
attributes :host, :type, :password
```

lzap replied:
Interesting fields for spice can be found here: http://linux.die.net/man/1/virt-install

But I also do not have names that fog uses.

ekohl commented:
I guess for hammer to use it, you'd need to un-websocket the connection.

I have been playing with the thought to raise 422 for non-VNC until it is implemented. V2 API is experimental anyway.

domcleal commented:
Yeah, though it could equally be advantageous to still proxy traffic via the Foreman server so you don't need direct connectivity from clients to the hypervisors.  I suppose the main downside is that no VNC client is going to know the WebSockets protocol, so Hammer would probably have to proxy it again... @mbacovsky?

The "log" console should be trivial, and perhaps someone (@lzap?) could check what's needed for SPICE consoles on libvirt and we can add it in here.

domcleal commented:
Actually, probably the best place to look for SPICE is https://github.com/theforeman/foreman/blob/develop/app/helpers/compute_resources_vms_helper.rb#L33

ekohl commented:
Initial update that only adds VNC. I'd like to add tests as well. What is needed to get this into a mergeable shape?

domcleal commented:
@ekohl implementations for all three console classes, preferably tests, and a clear indication of whether this is usable for hammer or not, or if we need to do some further work (ping @mbacovsky @tstrachota).

dLobatog commented:
I'm not against the necessity of having a direct connection from the client to the hypervisors, I'm not a hammer expert but I can't see why this wouldn't work through it. Possibly it'll be tricky to set it up for the interactive mode. @ekohl , do you have some free time to work on what @domcleal mentioned above and merge so Jenkins can run? 

I can't see a reason not to have this in even if it's just VNC, and return some message "SPICE through API is not implemented" until we figure it out. On the same vein, we could possibly think of returning the URL to the console when returning the OpenStruct is not possible, although I know this is technically questionable (2 different kind of responses for a same API call) :laughing:

ekohl commented:
@eLobato currently time is not widely available, so no promises. I'll at least rebase it now.

Fun fact: you can do a pull request to a pull request. Since it's just a branch, you can create a PR in my repo for that branch and when I merge it, this PR is updated.

dLobatog commented:
@ekohl heh I know, I've been using that in other projects. Tests won't pass until you define a permission for the method 'console' in api/v2/hosts.

ohadlevy commented:
@ekohl ping?

ekohl commented:
Rebased and added the permission mapping as @eLobato suggested. Note that since I changed employers, my use case for this is gone. Any help with finishing it would be appreciated.

ohadlevy commented:
Would be really nice to have hammer plugin to run vnc for you

ekohl commented:
Now API v2 is no longer experimental, I feel there should be a proper API definition first. I posted https://groups.google.com/forum/#!searchin/foreman-dev/console$20api/foreman-dev/PT-5RFqr0nw/ixYBOe75G60J back in January, but had no response.

dLobatog commented:
@ekohl That definition looks good to me

```
"console": {
    "host": "foreman.example.org",
    "port": 5900,
    "password": "something",
    "type": "vnc"
} 
```

We can start with that and add new options as users require them, for instance certificates, spice-only stuff, etc... Do you have any time to work on this lately? It's nearly done except for tests from what I'm seeing here.

ekohl commented:
@eLobato not that much time, but I can try. But is there a convention for experimental APIs now v2 is stable?

dLobatog commented:
@ekohl ok to close? It's been WIP for a long time :smile:

ekohl commented:
@eLobato personally I have no more use case for this so I think it's OK to close. Still an interesting feature though so hopefully it's a good starting point for others :)
